### ISSUE TITLE
[TreeView] New properties for Selected State and Click event on the container area 

### ISSUE THREAD

**Issue by drakonux** [liferay/clay#4949]
# Context
The TreeView component was designed to cover many cases for Liferay products. So, by default, the component can't solve all the product cases. Instead, it needs to be set depending on the case.

This proposal concluded from a conversation in a Slack thread where two product teams were discussing how a node should behave on click (ExpandOnClick vs CheckOnClick). Initially, we agreed that both teams would use the default interaction, which could be a possible solution. But Lexicon sees benefit for the experience if we invest in additional development to solve the following cases. 

## Checked vs. Selected
If the TreeView is using the Single Selection mode. A node can be selected but not checked. 
If the TreeView is using one of the Multiple Selection modes. A node can be selected and/or checked. 
 
![Select vs Checked](https://user-images.githubusercontent.com/21216422/175260213-0ca439e7-3935-4390-b58d-683509bf5af8.png)

For example, the Selected state makes sense for the Single Selection mode. 

![Single Selection](https://user-images.githubusercontent.com/21216422/175309486-51329dbe-a964-4086-bfc4-24d9c74daab2.png)

But it doesn't work for the Multiple Selection mode except if it's used for a master-detail pattern — Disclaimer; product teams haven't proposed this case.

![Multiple Selection - Master-Detail](https://user-images.githubusercontent.com/21216422/175309523-1da81ff8-3547-4668-a118-148f2db3dbc8.png)

Thus, when the Multiple Selection mode is set, the hover state of a node makes no sense if it triggers the Selected state because, in most cases, the TreeView is not used to apply a Master-Detail pattern. 

![Multiple Selection](https://user-images.githubusercontent.com/21216422/175309570-bd5dc39d-49eb-4a95-b90e-a14db6fda68a.png)

So for the no Master-Detail pattern cases on Multiple Selection, it's weird to see a node selected because the hover state highlights the node but clicking it selects the node, and that's it. So if we prevent the Select state for nodes, we must also avoid the Hover state or **manage the Click event**. If-else we are tricking the user because nothing happens by interacting with the node.  

## Types of Nodes

As a reminder, we proposed the following types of nodes:

- Container: a parent node that can't be selected **nor checked**.
- Parent: a node with one or more child nodes. It can be expanded (open) or collapsed (close).
- End: a node that does not have child nodes; an end node may be either a root node or a child node. 

# Proposal

- Allow turning off the Selected state of nodes
  - Container nodes 
  - Multiple Selection for no Master-Detail cases
- Allow product teams to manage the Click interaction on the container area.

The following cases describe how product teams will set the TreeView in DXP interacting with the container area:

## For Single Selection 

![Single Selection - Container Nodes](https://user-images.githubusercontent.com/21216422/175299965-136464bf-3fe9-4812-9841-aa9296c09ba1.png)

1. On Click, the container node is expanded/collapsed. 
2. On Double Click, the container node is expanded/collapsed

**Note**: A couple of DXP teams will use the double click interaction to edit the node's name. Thus, the current `expandDoubleClick` property will conflict with this new property once they request it.

![Single Selection - Parent Nodes](https://user-images.githubusercontent.com/21216422/175302632-6dc036f0-12ea-48b6-b4c8-3400d030840b.png)

1. On Click, the parent node changes to Selected.
2. On Double Click, the parent node changes to Selected and is expanded/collapsed.

![Single Selection - End Nodes](https://user-images.githubusercontent.com/21216422/175301912-26e0e37d-f61d-4650-9104-a5eb7e4673d3.png)

1. On Click, the end node changes to Selected. 
2. On Double Click, the end node changes to Selected. 

## For Multiple Selection

Double Click interaction **must not be provided** using the Multiple Selection mode to prevent interaction mistakes. 

![Multiple Selection - Container Nodes](https://user-images.githubusercontent.com/21216422/175307534-fab68081-4f25-433f-863e-f9d6d6b03699.png)

1. On Click, the container node is expanded/collapsed. 

![Multiple Selection - Parent Nodes](https://user-images.githubusercontent.com/21216422/175308477-4f1453ca-8453-4795-853c-9d8817d77218.png)

1. On Click, the parent node is checked/unchecked.

![Multiple Selection - End Nodes](https://user-images.githubusercontent.com/21216422/175309697-2eeb5aee-ff46-433f-9207-72fcdf02a4ab.png)

1. On Click, the end node is checked/unchecked.  in the future

# Alternative solution

The approach in this case. It's to not innovate. Thus, for the sake of interaction:

- The container area doesn't trigger the hover visual state if the node can't be selected (any node type).
- Expand/Collapse a node is triggered via the expander button.
- Check/Uncheck a node is triggered via the checkbox. 

![Single Selection - No Container Interaction](https://user-images.githubusercontent.com/21216422/175320167-550a61c8-b3b5-44bd-9939-793300942933.png)

Using the Multiple Selection for no Master-Detail cases should be the following:

![Multiple Selection - No Container Interaction](https://user-images.githubusercontent.com/21216422/175318924-918a4a29-2884-480a-ba68-32f44672062c.png)

cc\ @liferay-lexicon @claraizquierdo @susanavr @victorg1991 @carolrodmar

**Comment by matuzalemsteles** [comment]
Just to avoid confusion about selection, we don't have selection vs checked distinction, because the design part separates checked vs selection, for us both are the same thing at API level. Selection in this proposal has a divergent behavior so to try to reduce the confusion we can call it active. The cases of single selection and multiple selection regardless of whether the interaction is clicked on the checkbox or not, are selection behaviors.

> So for the no Master-Detail pattern cases on Multiple Selection, it's weird to see a node selected because the hover state highlights the node but clicking it selects the node, and that's it. So if we prevent the Select state for nodes, we must also avoid the Hover state or manage the Click event. If-else we are tricking the user because nothing happens by interacting with the node.

The current behavior is that clicking on the node will only expand, visually it only has the hover effect but the selection/active is not visible, this behavior is only visible on the single selection. This behavior can also change according to the previous proposal that some products can have the behavior of clicking on the node and also selecting the item but not the active state. Just to clear up the confusion, selecting is the same thing for us as checked/unchecked.

> Allow turning off the Selected state of nodes -> Container nodes

Well, we can say that this is possible today, when you don't render a `<Checkbox />` in the root node or as you call it container, it is not possible to select/checked via user interaction.

In the default behavior, multiple selection, the selection happens only via user interaction clicking on the checkbox or if the developer calls the API `selection.toggle(key)` in the node click event for example.

In single selection behavior, it selects any node type but developer can choose to ignore selection behavior for a specific node, this example here https://storybook.clayui.com/?path=/story/design-system-components-treeview--demo-categories-single it is selecting any node on click but can ignore selection.

```jsx
<TreeView.ItemStack
  onClick={(event) => {
    if (item.vocabulary) {
      event.preventDefault();
    }
  }}
>
  ...
</TreeView.ItemStack>
```

In other words, in the end, the TreeView component does not put specific rules for each behavior, it just has the APIs and composition in favor for the developer to configure according to the rules defined by the design and Lexicon.

> Allow product teams to manage the Click interaction on the container area.

This is also possible, as I commented above, to ignore the default behaviors of a Node, of selection, to add selection behaviors, and so on.

**Comment by drakonux** [comment]
> TreeView has no distinction of Node types, we did not implement this because it creates a much greater complexity to manage on our side and creates a very unnecessary abstraction, in composition the developer can create his node type rules according to his use case, for example, triggering the selection via the selection.toggle API, canceling more general default behaviors, when we start to include specific and conflicting behaviors it starts to create a very big complex and it becomes very difficult for us to manage and create something really flexible for all.

It's ok, and I agree. But these changes are crucial and must be shared with us when you decide on them because we document for product designers in product teams. Thus, even if you make a decision that is not aligned with the definition, we set previously. It would help if you communicated it because we must decide between keeping the inconsistency or aligning the docs with the development. 

> As I see it I'm not sure but you can confirm here, that the term, the selection is getting a little more confusing for me because it behaves differently in some places especially when we were migrating. Just my idea about it is that selection in a single selection is the same thing as checked but visually behaving differently. In this new proposal, this changes when in multiple selection mode, the term selection becomes something more related to just one click. For example, selection for us is the same thing as checked/uncheck at API level regardless of the selection mode, with these new behaviors for selection this changes everything because it becomes the state not of selection but of active.

This is not from this proposal. We defined that an item should be selectable and checkable a time ago (review the documentation). So probably the naming wasn't correct, but it was there.

The confusion is caused because selection for the single selection uses a "highlighted" state of the item (yes, calling it active could be better), and for multi-selection, we use the checkboxes to "select" or check items. So in a different way of saying it, an item is selected AND checked on a single selection. 

> I'm going to do some more testing with the component currently, configuring it to get to these use cases and try to understand how we can implement this.

If you confirm that product teams can define that behavior is enough for us and Clay doesn't have to implement new properties.

> The current behavior is that clicking on the node will only expand visually it only has the hover effect but the selection/active is not visible, this behavior is only visible on the single selection. 

This has to be a property because it depends on the use case. 

> This behavior can also change according to the previous proposal that some products can have the behavior of clicking on the node and also selecting the item but not the active state. Just to clear up the confusion, selecting is the same thing for us as checked/unchecked.

Well, as I said above it could be a case at some moment where the active state would be used on a multi-selection TreeView and we will have problems. Until that it's ok. 

As it's now as long as the active state is not set when a user interacts with an item on multi-selection (as I explained before, the hover states induce to click the item so product teams need to manage the event for something) if we provide a way to reach both conditions (no active state and check/expand on click) it's ok. 

> In single selection behavior, it selects any node type but developer can choose to ignore selection behavior for a specific node, this example here https://storybook.clayui.com/?path=/story/design-system-components-treeview--demo-categories-single it is selecting any node on click but can ignore selection.

Perfect then. 

> This is also possible, as I commented above, to ignore the default behaviors of a Node, of selection, to add selection behaviors, and so on.

Summarizing then :D the product teams can match the behaviors I described, aren't they?

**Comment by matuzalemsteles** [comment]
> It's ok, and I agree. But these changes are crucial and must be shared with us when you decide on them because we document for product designers in product teams. Thus, even if you make a decision that is not aligned with the definition, we set previously. It would help if you communicated it because we must decide between keeping the inconsistency or aligning the docs with the development.

For sure, but we have to understand that development and how people use components is different from design, we're not always going to be able to be something 1:1 at the definition level. For example, the design-level definition makes sense but for using the component it doesn't, I don't see it as inconsistent because the design defines the design level and development shows the developer how to achieve the design definition. We must not look at this as inconsistency!

At least this detail in respect, there was nothing defined at the definition level! This is something new right?

Just in relation to that, in my opinion, it makes sense to impose or add names to the nodes but at the development level the developer will impose these restrictions. The component only offers the ways to configure it for the use cases.

Because defining this at the component level or something via API would make things more complex and it wouldn't apply to all TreeView use cases, at least considering what we see in the documentation.

> This is not from this proposal. We defined that an item should be selectable and checkable a time ago (review the documentation). So probably the naming wasn't correct, but it was there.
> The confusion is caused because selection for the single selection uses a "highlighted" state of the item (yes, calling it active could be better), and for multi-selection, we use the checkboxes to "select" or check items. So in a different way of saying it, an item is selected AND checked on a single selection.

I believe it's this line https://liferay.atlassian.net/wiki/spaces/ENGLEXICON/pages/1973518798/Interaction#Selection maybe it wasn't clear that this behavior could be used regardless of the selection mode. I think the term used also ended up confusing things 🥲.

> This has to be a property because it depends on the use case.

So we want to be able to disable the hover state as well but would we leave that by default? @pat270 maybe I need you here to come up with something to remove this when adding a class but that would have to be at a specific node level rather than turning it off for everyone.

> As it's now as long as the active state is not set when a user interacts with an item on multi-selection (as I explained before, the hover states induce to click the item so product teams need to manage the event for something) if we provide a way to reach both conditions (no active state and check/expand on click) it's ok.

I see, the component manages to provide this, for the developer to add check and expand state by clicking on the node or switching between them if necessary. I think the only thing missing is the above case of being able to take the hover off.

> Summarizing then :D the product teams can match the behaviors I described, aren't they?

I believe that in the vast majority of these use cases yes, maybe we need to adapt one case or another, for example being able to remove the hover and I have to check the behavior of expanding and selecting with the use cases of double click in a single selection. I'll check it out today and should get back to you with some more info.

Just a doubt, is this proposal related to these examples in Storybook? If so, I will use this issue to update them with these new rules.

- [Categories Single](https://storybook.clayui.com/?path=/story/design-system-components-treeview--demo-categories-single)
- [Categories Multiple](https://storybook.clayui.com/?path=/story/design-system-components-treeview--demo-categories-multiple)
- [Documents Multiple](https://storybook.clayui.com/?path=/story/design-system-components-treeview--demo-documents-multiple)

**Comment by matuzalemsteles** [comment]
I did some tests to cover these use cases, with this we are going to add a new API to allow the developer to trigger the item expansion in the same way we do with the selection and a new property in the `<TreeView.Item />` that allows defining if the node is in an active state to cover the scenario when you want an active node in multiple selection as in the example. This will allow you to configure the cases above for this scenario. I can add a demo to this as well to demonstrate how to achieve these use cases. I will send a PR for that.

**Comment by drakonux** [comment]
> For sure, but we have to understand that development and how people use components is different from design, we're not always going to be able to be something 1:1 at the definition level. For example, the design-level definition makes sense but for using the component it doesn't, I don't see it as inconsistent because the design defines the design level and development shows the developer how to achieve the design definition. We must not look at this as inconsistency!

Ok, ok probably it's not an inconsistency but Lexicon has to be aware. Just increase the communication with Lexicon please. I'm not fully assigned to Lexicon for months ago so it's super important that Clay talks frequently about everything Clay is doing with the Lexicon team (not just me). 

> I believe it's this line https://liferay.atlassian.net/wiki/spaces/ENGLEXICON/pages/1973518798/Interaction#Selection maybe it  wasn't clear that this behavior could be used regardless of the selection mode. I think the term used also ended up confusing things.

That's it, it says "can" not "has" but it's true, it's confusing. But I think It should be a property, as `expandedOnLoad` and others. 

Lexicon has to be close to the implementation for things like this. Because we need to provide 1:1 components for Product Designers in the Figma library and we should not place things on the Lexicon site that can't be built using Clay. 

> Just a doubt, is this proposal related to these examples in Storybook? If so, I will use this issue to update them with these new rules.

It's totally related. Just one thing, the double click interaction has to be defined by each product team because in some use cases in DXP they will use the double click for editing the name of the node (it wasn't requested yet but they will)

> I believe that in the vast majority of these use cases yes, maybe we need to adapt one case or another, for example being able to remove the hover and I have to check the behavior of expanding and selecting with the use cases of double click in a single selection. I'll check it out today and should get back to you with some more info

I talked with Clara (Tango's dev) days ago and she said she was able to match this behavior so it's ok. But it's good if you check everything and polish the component. Thanks for the effort Matu.

**Comment by matuzalemsteles** [comment]
> I talked with Clara (Tango's dev) days ago and she said she was able to match this behavior so it's ok. But it's good if you check everything and polish the component.

Oh, that's good, submit a new PR #4958 that adds two new possibilities to reach more of these use cases. This will be available in the next release on Monday.

- New public `expand` API that lets you control node expansion/collapse
- New property for node to control `active` state

I also updated both demos to follow these recommendations:

- [Categories Single](https://62bdf428a9cf07000aa9e275--next-storybook-clayui.netlify.app/?path=/story/design-system-components-treeview--demo-categories-single)
- [Categories Multiple](https://62bdf428a9cf07000aa9e275--next-storybook-clayui.netlify.app/?path=/story/design-system-components-treeview--demo-categories-multiple)

Also adding how to do this in the documentation, I still have to add a new property to disable the item hover.

**Comment by matuzalemsteles** [comment]
> I still have to add a new property to disable the item hover

Merging today PR #4964 which adds the new property to handle this. I'm going ahead and closing this PR, we've added new public APIs and new settings to handle these scenarios.

- #4958
- #4964

The demos have been updated to follow these recommendations and demonstrate how the TreeView can be configured.
